device/src/secure_boot.rs

```diff
@@ -0,0 +1,382 @@
+        let hash = compute_sha256_hardware(sha, &input);
+        mask.extend_from_slice(&hash);
+        counter += 1;
+    }
+
+    mask.truncate(mask_len);
+    mask
+}
+
+// Helper function to compute SHA256 using hardware peripheral
+fn compute_sha256_hardware(sha: &mut Sha, data: &[u8]) -> [u8; 32] {
+    let mut hasher = sha.start::<Sha256>();
+
+    let mut remaining = data;
+    while !remaining.is_empty() {
+        remaining = nb::block!(hasher.update(remaining)).unwrap();
+    }
+
+    let mut result = [0u8; 32];
+    nb::block!(hasher.finish(&mut result)).unwrap();
+    result
+}
+
+// Find secure boot key digest from eFuse by checking KEY_PURPOSE fields
+fn find_secure_boot_key() -> Option<[u8; 32]> {
+    use esp_hal::efuse::{
+        KEY0, KEY1, KEY2, KEY3, KEY4, KEY5, KEY_PURPOSE_0, KEY_PURPOSE_1, KEY_PURPOSE_2,
+        KEY_PURPOSE_3, KEY_PURPOSE_4, KEY_PURPOSE_5, SECURE_BOOT_KEY_REVOKE0,
+        SECURE_BOOT_KEY_REVOKE1, SECURE_BOOT_KEY_REVOKE2,
+    };
+    // Key purpose values and their revoke fields (from ESP32-C3 TRM Table 4.3-1 & 4.3-2)
+    let secure_boot_digests = [
+        (9, SECURE_BOOT_KEY_REVOKE0),  // DIGEST0
+        (10, SECURE_BOOT_KEY_REVOKE1), // DIGEST1
+        (11, SECURE_BOOT_KEY_REVOKE2), // DIGEST2
+    ];
+
+    let key_purpose_fields = [
+        KEY_PURPOSE_0,
+        KEY_PURPOSE_1,
+        KEY_PURPOSE_2,
+        KEY_PURPOSE_3,
+        KEY_PURPOSE_4,
+        KEY_PURPOSE_5,
+    ];
+    let key_data_fields = [KEY0, KEY1, KEY2, KEY3, KEY4, KEY5];
+
+    // Search through all key blocks
+    for (i, &purpose_field) in key_purpose_fields.iter().enumerate() {
+        let purpose: u8 = Efuse::read_field_le(purpose_field);
+
+        // Find matching secure boot digest revoke field
+        if let Some((_, revoke_field)) = secure_boot_digests
+            .iter()
+            .find(|(purpose_val, _)| *purpose_val == purpose)
+        {
+            // Check if this key is revoked
+            let is_revoked = Efuse::read_bit(*revoke_field);
+            if !is_revoked {
+                // Read the key data (32 bytes)
+                let key_data: [u8; 32] = Efuse::read_field_le(key_data_fields[i]);
+                return Some(key_data);
+            }
+        }
+    }
+    None
+}
+
+/// Check if secure boot is enabled by looking for secure boot key digests in eFuse
+pub fn is_secure_boot_enabled() -> bool {
+    find_secure_boot_key().is_some()
+}
+
+pub fn verify_secure_boot<'a, S>(
+    app_partition: &FlashPartition<S>,
+    rsa: &mut Rsa<'_, Blocking>,
+    sha: &mut Sha,
+) -> Result<(), SecureBootError<'a>>
+where
+    S: ReadNorFlash + embedded_storage::nor_flash::NorFlash,
+{
+    let mut signature_block = vec![0x00; SECTOR_SIZE];
+    let mut signature_found = false;
+    let mut signature_sector_index = 0u32;
+
+    // Find signature block in app partition using magic bytes (magic byte, secure boot v2, padding)
+    for i in 0..app_partition.n_sectors() {
```

Comment:
```suggestion
    for i in 0..app_partition.n_sectors().rev() {
```

iterate backwards so we don't have to have this crazy comment below?